src/Form/MediaLibraryEmbeddableForm.php

```diff
@@ -0,0 +1,171 @@
+        '@providers' => implode(', ', $providers),
+      ]),
+      '#required' => FALSE,
+      '#attributes' => [
+        'placeholder' => 'https://',
+      ],
+    ];
+
+    if ($authorized_for_unstructured) {
+      $form['container']['field_media_embeddable_code'] = [
+        '#type' => 'textarea',
+        '#title' => $this->t('Embed Code'),
+        '#description' => $this->t('Use this field to paste in embed codes which are not available through oEmbed'),
+        '#required' => FALSE,
+        '#attributes' => [
+          'placeholder' => '',
+        ],
+      ];
+    }
+
+    $ajax_query = $this->getMediaLibraryState($form_state)->all();
+    $ajax_query += [FormBuilderInterface::AJAX_FORM_REQUEST => TRUE];
+    $form['container']['submit'] = [
+      '#type' => 'submit',
+      '#value' => $this->t('Add'),
+      '#button_type' => 'primary',
+      '#validate' => ['::validateEmbeddable'],
+      '#submit' => ['::addButtonSubmit'],
+      '#ajax' => [
+        'callback' => '::updateFormCallback',
+        'wrapper' => 'media-library-wrapper',
+        // Add a fixed URL to post the form since AJAX forms are automatically
+        // posted to <current> instead of $form['#action'].
+        // @todo Remove when https://www.drupal.org/project/drupal/issues/2504115
+        //   is fixed.
+        'url' => Url::fromRoute('media_library.ui'),
+        'options' => ['query' => $ajax_query],
+      ],
+    ];
+
+    return $form;
+  }
+
+  /**
+   * Informs us if we are working with an unstructured embed.
+   *
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current form state.
+   * @return bool
+   *   True if unstructured, otherwise false.
+   */
+  public function isUnstructured(FormStateInterface $form_state) {
+    $embed_code = $form_state->getValue('field_media_embeddable_code');
+    if ($embed_code) {
+      return TRUE;
+    }
+    return FALSE;
+  }
+
+  /**
+   * Validates the oEmbed URL.
+   *
+   * @param array $form
+   *   The complete form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current form state.
+   */
+  public function validateEmbeddable(array &$form, FormStateInterface $form_state) {
+    // No validation necessary if we have an embed code.
+    if ($this->isUnstructured($form_state)) {
+      return;
+    }
+    parent::validateUrl($form, $form_state);
+  }
+
+  /**
+   * Submit handler for the add button.
+   *
+   * @param array $form
+   *   The form render array.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   */
+  public function addButtonSubmit(array $form, FormStateInterface $form_state) {
+    if ($this->isUnstructured($form_state)) {
+      $values = [$form_state->getValue('field_media_embeddable_code')];
+    } else {
+      $values = [$form_state->getValue('field_media_embeddable_oembed')];
```

Comment:
> field names should be avoided in the class. if the field name changes, everything will break.

Ah, do you have an example of how to avoid this? How'd you do it for the other media?

Reply:
This has been fixed throughout the module.